x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheService.java

```diff
@@ -52,31 +82,63 @@
         Setting.Property.NodeScope
     );
 
+    public static final TimeValue MIN_SNAPSHOT_CACHE_SYNC_INTERVAL = TimeValue.timeValueSeconds(10L);
+    public static final Setting<TimeValue> SNAPSHOT_CACHE_SYNC_INTERVAL_SETTING = Setting.timeSetting(
+        SETTINGS_PREFIX + "sync_interval",
+        TimeValue.timeValueSeconds(60L),                        // default
+        MIN_SNAPSHOT_CACHE_SYNC_INTERVAL,                       // min
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+
+    private static final Logger logger = LogManager.getLogger(CacheService.class);
+
+    private final ClusterService clusterService;
+    private final NodeEnvironment nodeEnvironment;
+    private final ThreadPool threadPool;
+    private final CacheSynchronizationTask cacheSyncTask;
     private final Cache<CacheKey, CacheFile> cache;
     private final ByteSizeValue cacheSize;
     private final Runnable cacheCleaner;
     private final ByteSizeValue rangeSize;
 
-    public CacheService(final Runnable cacheCleaner, final Settings settings) {
-        this(cacheCleaner, SNAPSHOT_CACHE_SIZE_SETTING.get(settings), SNAPSHOT_CACHE_RANGE_SIZE_SETTING.get(settings));
-    }
-
-    // exposed for tests
-    public CacheService(final Runnable cacheCleaner, final ByteSizeValue cacheSize, final ByteSizeValue rangeSize) {
-        this.cacheSize = Objects.requireNonNull(cacheSize);
+    public CacheService(
+        final Settings settings,
+        final ClusterService clusterService,
+        final ThreadPool threadPool,
+        final NodeEnvironment nodeEnvironment,
+        final Runnable cacheCleaner
+    ) {
+        this.clusterService = Objects.requireNonNull(clusterService);
+        this.nodeEnvironment = Objects.requireNonNull(nodeEnvironment);
+        this.threadPool = Objects.requireNonNull(threadPool);
+        this.cacheSize = SNAPSHOT_CACHE_SIZE_SETTING.get(settings);
         this.cacheCleaner = Objects.requireNonNull(cacheCleaner);
-        this.rangeSize = Objects.requireNonNull(rangeSize);
+        this.rangeSize = SNAPSHOT_CACHE_RANGE_SIZE_SETTING.get(settings);
         this.cache = CacheBuilder.<CacheKey, CacheFile>builder()
             .setMaximumWeight(cacheSize.getBytes())
             .weigher((key, entry) -> entry.getLength())
             // NORELEASE This does not immediately free space on disk, as cache file are only deleted when all index inputs
             // are done with reading/writing the cache file
             .removalListener(notification -> IOUtils.closeWhileHandlingException(() -> notification.getValue().startEviction()))
             .build();
+
+        if (DiscoveryNode.isDataNode(settings)) {
```

Comment:
++ I think not instantiating it would be nice. 
Mild worry on this :) <- it might be a little cumbersome to get it right due to transport actions? (haven't checked this here in detail but I remember trying to do a similar thing elsewhere and it turned out to be tricky)